File: src/main/java/com/f2prateek/bee/SchedulersHook.java
package com.f2prateek.bee;

import java.util.concurrent.atomic.AtomicReference;
import rx.Scheduler;
import rx.schedulers.Schedulers;

/**
 * Too lazy to setup proper DI for this project.
 * Waiting on https://github.com/ReactiveX/RxJava/issues/2297.
 */
public final class SchedulersHook {
  private SchedulersHook() {
    throw new AssertionError("no instances.");
  }

  private static final AtomicReference<Scheduler> COMPUTATION = new AtomicReference<>();

  public static Scheduler computation() {
    if (COMPUTATION.get() == null) {
      COMPUTATION.compareAndSet(null, Schedulers.computation());
      // Call get() again in case of thread-race so the winner will always get returned.
    }
    return COMPUTATION.get();
  }

  public static void setComputationScheduler(Scheduler scheduler) {
    if (!COMPUTATION.compareAndSet(null, scheduler)) {
      throw new IllegalStateException("Scheduler was already registered: " + COMPUTATION.get());
    }
  }
}

File: src/main/java/com/f2prateek/bee/SchedulersHook.kt
package com.f2prateek.bee

import rx.Scheduler
import rx.schedulers.Schedulers
import java.util.concurrent.atomic.AtomicReference

/**
 * Too lazy to setup proper DI for this project.
 * Waiting on https://github.com/ReactiveX/RxJava/issues/2297.
 */
class SchedulersHook private constructor() {
    init {
        throw AssertionError("no instances.")
    }

    companion object {
        private val COMPUTATION = AtomicReference<Scheduler>()

        fun computation(): Scheduler {
            if (COMPUTATION.get() == null) {
                COMPUTATION.compareAndSet(null, Schedulers.computation())
                // Call get() again in case of thread-race so the winner will always get returned.
            }
            return COMPUTATION.get()
        }

        fun setComputationScheduler(scheduler: Scheduler) {
            if (!COMPUTATION.compareAndSet(null, scheduler)) {
                throw IllegalStateException("Scheduler was already registered: " + COMPUTATION.get())
            }
        }
    }
}
